### PR TITLE
rpc: Add feeRate argument to bumpFee RPC

### LINKED ISSUE #16203: allow feerate argument in `bumpfee` in addition to confTarget
If someone has their own home-grown fee estimator, let's allow them to use it on command line.

### PR DESCRIPTION
**Pull request by ezegom**
Fixes #16203. 

On PR #15996,  the totalfee argument in `bumpfee` gets deprecated. The feeRate argument in `bumpfee` serves as a replacement to totalFee. 

Small changes to `CreateRateBumpTransaction`
- If the coin_control passed in has a feeRate set then we check if the feeRate is OK and if so, create the bump transaction. The logic to verify if the feeRate is good should be identical  to the logic that verifies the totalFee value.
- If the coin_control does not have a feeRate set then we estimate the fee as we used to.

### CODE REVIEW

**Comment by instagibbs** [comment]
concept ACK

File: src/wallet/feebumper.cpp [hunk 1]
```diff
@@ -57,6 +57,58 @@ static feebumper::Result PreconditionChecks(interfaces::Chain::Lock& locked_chai
     return feebumper::Result::OK;
 }
 
+//! Check if the user provided a valid feeRate 
+static feebumper::Result CheckFeeRate(const CWallet* wallet, const CWalletTx& wtx, const CFeeRate& newFeerate, const int64_t maxTxSize, std::vector<std::string>& errors) {
+    // check that fee rate is higher than mempool's minimum fee
```
Comment (instagibbs) [inline src/wallet/feebumper.cpp]: is this block of code cribbed from somewhere to compare with?

File: src/wallet/feebumper.cpp [hunk 2]
```diff
@@ -57,6 +57,58 @@ static feebumper::Result PreconditionChecks(interfaces::Chain::Lock& locked_chai
...
+    
+    // Given old total fee and transaction size, calculate the old feeRate
+    CAmount old_fee = wtx.GetDebit(ISMINE_SPENDABLE) - wtx.tx->GetValueOut();
+    const int64_t txSize = GetVirtualTransactionSize(*(wtx.tx));    
+    CFeeRate nOldFeeRate(old_fee, txSize);
+    // Min total fee is old fee + relay fee
+    CAmount minTotalFee = nOldFeeRate.GetFee(maxTxSize) + incrementalRelayFee.GetFee(maxTxSize);
+
+    if (new_total_fee < minTotalFee) {
+        errors.push_back(strprintf("Insufficient totalFee %s, must be at least %s (oldFee %s + incrementalFee %s)",
```
Comment (instagibbs) [inline src/wallet/feebumper.cpp]: `totalFee` isn't a thing

File: src/wallet/feebumper.cpp (hunk 1 above)
Comment (ezegom) [reply src/wallet/feebumper.cpp]: Yeah, this code is from the `CreateTotalBumpTransaction` method https://github.com/bitcoin/bitcoin/blob/b89249e0de8ad1848d487dd3e7a85a0d6d456de8/src/wallet/feebumper.cpp#L199-L247

File: src/wallet/feebumper.cpp (hunk 2 above)
Comment (ezegom) [reply src/wallet/feebumper.cpp]: > I'll rewrite these error messages.

**Comment by promag** [comment]
Concept ACK. Some notes:

 - try to keep commits organized
 - please see git log and pick one of the commit message format

**Comment by ezegom** [comment]
> Concept ACK. Some notes:
> 
> * try to keep commits organized
> * please see git log and pick one of the commit message format

I'll change the commits from "<rpc>" to "rpc:". My bad, should've taken a look at the git log before assuming this style. 
Regarding the organization, what do you feel is not organized?

**Comment by MarcoFalke** [comment]
Could rebase on master and modify the release notes in the section "removed or deprecated rpcs"?

File: test/functional/wallet_bumpfee.py [hunk 3]
```diff
@@ -176,7 +188,6 @@ def test_bumpfee_with_descendant_fails(rbf_node, rbf_node_address, dest_address)
     rbf_node.sendrawtransaction(tx["hex"])
     assert_raises_rpc_error(-8, "Transaction has descendants in the wallet", rbf_node.bumpfee, parent_id)
 
-
```
Comment (Sjors) [inline test/functional/wallet_bumpfee.py]: Nit: avoid dropping or adding whitespace in code that you're not touching.

**Review by Sjors — CHANGES_REQUESTED** [review]
Concept ACK. 

In the description, if you replace "Draft commit for issue" with "Fixes" then Github will automatically close that issue when this PR is merged, which saves maintainers some effort.

> Regarding the organization, what do you feel is not organized?

You should use git interactive rebase to reduce the number of commits, for example: 

* commit `linter: fix whitespace`: add this into the commit that added the incorrect whitespace (`fixup`)
* commits `rpc: test bumpfee feeRate argument`: squash these together
* commit `doc: Add release note...`: drop this commit, because #16504

This makes it easier to review a pull request one commit at a time.

* commit `Changes the verbosity of msbuild`: move to separate pull request

File: test/functional/wallet_bumpfee.py (hunk 3 above)
Comment (ezegom) [reply test/functional/wallet_bumpfee.py]: fixed

**Comment by ezegom** [comment]
Thanks @Sjors, 
I cleaned up the commits accordingly.

File: doc/release-notes.md [hunk 4]
```diff
@@ -92,7 +92,9 @@ Deprecated or removed RPCs
 --------------------------
 
 - The `totalFee` option of the `bumpfee` RPC has been deprecated and will be
-  removed in 0.20.  To continue using this option start with
+  removed in 0.20. The `feeRate` option is now the preferred way to specify
+  the fee for the bump transaction.
```
Comment (Sjors) [inline doc/release-notes.md]: Nit: bumped

File: src/wallet/rpcwallet.cpp [hunk 5]
```diff
@@ -3361,7 +3361,7 @@ static UniValue bumpfee(const JSONRPCRequest& request)
             if (fee_rate <= 0) {
                 throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("Invalid feeRate %s (must be greater than 0)", fee_rate.ToString()));
             }
-            coin_control.m_feerate = fee_rate; 
+            coin_control.m_feerate = fee_rate;
```
Comment (Sjors) [inline src/wallet/rpcwallet.cpp]: Nit: move whitespace fix to commit where the whitespace mistake was made

File: src/wallet/rpcwallet.cpp [hunk 6]
```diff
@@ -3278,7 +3278,7 @@ static UniValue bumpfee(const JSONRPCRequest& request)
                 "All inputs in the original transaction will be included in the replacement transaction.\n"
                 "The command will fail if the wallet or mempool contains a transaction that spends one of T's outputs.\n"
                 "By default, the new fee will be calculated automatically using estimatesmartfee.\n"
-                "The user can specify a confirmation target for estimatesmartfee.\n"
```
Comment (Sjors) [inline src/wallet/rpcwallet.cpp]: Nit: don't remove whitespace here

File: src/wallet/feebumper.cpp [hunk 7]
```diff
@@ -304,7 +361,8 @@ Result CreateRateBumpTransaction(CWallet* wallet, const uint256& txid, const CCo
     return Result::OK;
 }
 
-bool SignTransaction(CWallet* wallet, CMutableTransaction& mtx) {
```
Comment (Sjors) [inline src/wallet/feebumper.cpp]: Nit: you're not touching this function, so don't fix the brackets please

File: src/wallet/feebumper.cpp [hunk 8]
```diff
@@ -4,16 +4,16 @@
 
 #include <consensus/validation.h>
 #include <interfaces/chain.h>
-#include <wallet/coincontrol.h>
```
Comment (Sjors) [inline src/wallet/feebumper.cpp]: nit: don't fix the include order; that's best done when you add a new include

File: src/wallet/feebumper.cpp [hunk 9]
```diff
@@ -71,8 +152,7 @@ bool TransactionCanBeBumped(const CWallet* wallet, const uint256& txid)
     return res == feebumper::Result::OK;
 }
 
-Result CreateTotalBumpTransaction(const CWallet* wallet, const uint256& txid, const CCoinControl& coin_control, CAmount total_fee, std::vector<std::string>& errors,
-                                  CAmount& old_fee, CAmount& new_fee, CMutableTransaction& mtx)
+Result CreateTotalBumpTransaction(const CWallet* wallet, const uint256& txid, const CCoinControl& coin_control, CAmount total_fee, std::vector<std::string>& errors, CAmount& old_fee, CAmount& new_fee, CMutableTransaction& mtx)
```
Comment (Sjors) [inline src/wallet/feebumper.cpp]: nit: don't fix things you're not otherwise touching

File: src/wallet/feebumper.cpp [hunk 10]
```diff
@@ -142,12 +222,12 @@ Result CreateTotalBumpTransaction(const CWallet* wallet, const uint256& txid, co
     }
 
     // Check that in all cases the new fee doesn't violate maxTxFee
-     const CAmount max_tx_fee = wallet->m_default_max_tx_fee;
-     if (new_fee > max_tx_fee) {
-         errors.push_back(strprintf("Specified or calculated fee %s is too high (cannot be higher than -maxtxfee %s)",
-                               FormatMoney(new_fee), FormatMoney(max_tx_fee)));
-         return Result::WALLET_ERROR;
-     }
+    const CAmount max_tx_fee = wallet->m_default_max_tx_fee;
```
Comment (Sjors) [inline src/wallet/feebumper.cpp]: nit: don't fix this

**Review by Sjors — COMMENTED** [review]
Thanks for condensing the commits.

You can probably remove the WIP and Draft status; those are meant to prevent too detailed review, but it's too late for that :-) 

There's still quite a bit of duplicate code between `CheckFeeRate` and `CreateTotalBumpTransaction`. Can you add an additional commit, similar to `rpc: bumpfee move feeRate estimation logic into sepparate method`, where you move this to a helper function? I'll review the actual functionality in more detail after that.

There's a couple of places where you or your editor fixed stuff in code that you're not otherwise touching. This increases review burden, because it's harder to see what changes are relevant, so is best avoided. I left a bunch of inline comments for that.

Couple more nits:
* commit `rpc: test bumpfee feeRate argument` should be named `[test] rpc bumpfee feeRate argument`

File: doc/release-notes.md (hunk 4 above)
Comment (ezegom) [reply doc/release-notes.md]: I initially wrote it as 'bumped' but went with 'bump'. My thought was that the bumped transaction is the one we are replacing, and the bump transaction is the new one?

Comment (promag) [reply doc/release-notes.md]: `the fee for the new transaction.`

Comment (ezegom) [reply doc/release-notes.md]: thats better

File: src/wallet/rpcwallet.cpp (hunk 5 above)
Comment (ezegom) [reply src/wallet/rpcwallet.cpp]: fixed

Comment (ezegom) [reply src/wallet/rpcwallet.cpp]: fixed

File: src/wallet/rpcwallet.cpp (hunk 6 above)
Comment (ezegom) [reply src/wallet/rpcwallet.cpp]: fixed

File: src/wallet/feebumper.cpp (hunk 7 above)
Comment (ezegom) [reply src/wallet/feebumper.cpp]: fixed :)

File: src/wallet/feebumper.cpp (hunk 8 above)
Comment (ezegom) [reply src/wallet/feebumper.cpp]: fixed

File: src/wallet/feebumper.cpp (hunk 9 above)
Comment (ezegom) [reply src/wallet/feebumper.cpp]: fixed this too

File: src/wallet/feebumper.cpp (hunk 10 above)
Comment (ezegom) [reply src/wallet/feebumper.cpp]: done

**Comment by ezegom** [comment]
@Sjors I should be done with your comments, let me know if there is anything missing. Only thing missing is to get rid of the code re-use on CreateTotalTransaction.

File: src/wallet/rpcwallet.cpp [hunk 11]
```diff
@@ -3332,13 +3335,15 @@ static UniValue bumpfee(const JSONRPCRequest& request)
...
                 {"estimate_mode", UniValueType(UniValue::VSTR)},
             },
             true, true);
-
-        if (options.exists("confTarget") && options.exists("totalFee")) {
-            throw JSONRPCError(RPC_INVALID_PARAMETER, "confTarget and totalFee options should not both be set. Please provide either a confirmation target for fee estimation or an explicit total fee for the transaction.");
+        if (options.exists("confTarget") && (options.exists("totalFee") || options.exists("feeRate"))) {
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "confTarget can't be set with totalFee or feeRate. Please provide either a conftarget for fee estimation, or an explicit total fee or fee rate for the transaction.");
+        } else if (options.exists("feeRate") && options.exists("totalFee")) {
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "feeRate can't be set along with totalFee. Please provide either a total fee or a fee rate (in satoshis per K) for the transaction.");
```
Comment (instagibbs) [inline src/wallet/rpcwallet.cpp]: Where else are we using satoshis per Kb? Seems like we're using BTC/kB everywhere else unless I'm missing a use?

Comment (ezegom) [reply src/wallet/rpcwallet.cpp]: You are correct, I used Satoshis per kB on my first iteration of this feature and forgot to change the error message. nice catch

**Comment by instagibbs** [comment]
(couldn't reach you on IRC) still working on this?

**Comment by fanquake** [comment]
Taken over in #16727.

**Pull request closed without merging**